dart/tool/java/generate_java.dart

```diff
@@ -1218,8 +1218,16 @@ class TypeRef {
           writer.addLine('return new $name(child);');
         } else {
           if (optional) {
-            writer.addLine('return json.get("$propertyName") == null ? '
-                'null : new $name((JsonObject) json.get("$propertyName"));');
+            writer.addLine('JsonObject obj = (JsonObject) json.get("$propertyName");');
```

Comment:
This looks reasonable; it’ll convert Null InstanceRef and Instance results to `null` when the expected result type is not InstanceRef or Instance.